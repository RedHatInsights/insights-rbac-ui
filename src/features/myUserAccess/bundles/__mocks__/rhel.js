import React from 'react';

const InsightsBundle = () => <div id="rhel-mock">This is just a mock insights bundle module</div>;

export default InsightsBundle;
