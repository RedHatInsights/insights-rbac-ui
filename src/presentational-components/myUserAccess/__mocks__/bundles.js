export const bundleData = [
  {
    entitlement: 'entitled',
    title: 'Entitled',
    apps: {
      eApp1: '/eapp1',
      eApp2: '/eapp2',
    },
    appsIds: ['a1', 'a2'],
  },
  {
    entitlement: 'not-entitled',
    title: 'Not entitled',
    apps: {
      neApp1: '/neapp1',
      neApp2: '/neapp2',
    },
    appsIds: ['na1', 'na2'],
  },
];
