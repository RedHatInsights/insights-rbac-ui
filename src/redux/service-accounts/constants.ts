export const NO_DATA = 'no-data';
export const LAST_PAGE = 'last-page';
export const RESULTS = 'results';
export const ERROR = 'error';
