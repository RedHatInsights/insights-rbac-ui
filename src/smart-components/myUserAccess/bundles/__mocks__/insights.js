import React from 'react';

const InsightsBundle = () => <div id="insights-mock">This is just a mock insights bundle module</div>;

export default InsightsBundle;
