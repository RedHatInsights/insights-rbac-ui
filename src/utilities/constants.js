export const RBAC_API_BASE = `/api/rbac/v1`;
