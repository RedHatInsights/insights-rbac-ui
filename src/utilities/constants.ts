export const RBAC_API_BASE = `/api/rbac/v1` as const;
export const RBAC_API_BASE_2 = `/api/rbac/v2` as const;
export const COST_API_BASE = `/api/cost-management/v1` as const;
export const INVENTORY_API_BASE = `/api/inventory/v1` as const;

export const DEFAULT_MUA_BUNDLE = 'rhel' as const;

export const DEFAULT_ACCESS_GROUP_ID = 'default-access' as const;
