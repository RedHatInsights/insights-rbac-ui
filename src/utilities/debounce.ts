import awesomeDebouncePromise from 'awesome-debounce-promise';
import type { AwesomeDebounceOptions } from 'awesome-debounce-promise';
import lodashDebounce from 'lodash/debounce';
import type { DebounceSettings } from 'lodash';

/**
 * Default debounce delay in milliseconds
 * 250ms is the sweet spot - fast enough to feel responsive, slow enough to reduce API calls
 */
export const DEFAULT_DEBOUNCE_DELAY = 250;

/**
 * Debounces an async function with race condition prevention
 *
 * @param fn - The async function to debounce
 * @param wait - Delay in milliseconds
 * @param options - Must include onlyResolvesLast: true for race-safe debouncing
 * @returns Debounced version of the function with cancel() method
 *
 * @example
 * const safeFetch = debounce(fetchId, 300, { onlyResolvesLast: true });
 */
export function debounce<T extends (...args: any[]) => Promise<any>>(
  fn: T,
  wait: number,
  options: AwesomeDebounceOptions & { onlyResolvesLast: true },
): T & { cancel(): void };

/**
 * Debounces a function (sync or async without race guards)
 *
 * @param fn - The function to debounce (sync or async)
 * @param wait - Delay in milliseconds (default: 250ms)
 * @param options - Standard lodash debounce options
 * @returns Debounced version of the function with cancel() and flush() methods
 *
 * @example
 * // Async functions
 * const debouncedSearch = debounce(
 *   async (query: string) => fetchResults(query)
 * );
 *
 * // Redux dispatch
 * const debouncedFetch = debounce(fetchData);
 *
 * // Cleanup on unmount
 * useEffect(() => () => debouncedFetch.cancel(), []);
 */
export function debounce<T extends (...args: any[]) => any>(
  fn: T,
  wait?: number,
  options?: DebounceSettings,
): T & { cancel(): void; flush(): ReturnType<T> | undefined };

// Implementation
export function debounce(fn: any, wait = DEFAULT_DEBOUNCE_DELAY, options: any = {}) {
  if ('onlyResolvesLast' in options && options.onlyResolvesLast === true) {
    return awesomeDebouncePromise(fn, wait, options);
  }
  return lodashDebounce(fn, wait, options);
}

/**
 * Legacy debounceAsync function - wraps awesome-debounce-promise for backwards compatibility
 *
 * @param asyncFunction - The async function to debounce
 * @param debounceTime - Delay in milliseconds (default: 250ms)
 * @param options - Awesome debounce options
 * @returns Debounced version of the function
 *
 * @deprecated Use debounce() with { onlyResolvesLast: true } when you need race-safe debouncing
 */
export function debounceAsync<T extends (...args: any[]) => Promise<any>>(
  asyncFunction: T,
  debounceTime = DEFAULT_DEBOUNCE_DELAY,
  options: Partial<AwesomeDebounceOptions> = { onlyResolvesLast: false },
): T & { cancel(): void } {
  return awesomeDebouncePromise(asyncFunction, debounceTime, options) as T & { cancel(): void };
}
