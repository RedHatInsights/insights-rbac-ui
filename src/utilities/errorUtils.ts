/**
 * Extracts user-friendly error messages from API error responses
 * Filters out technical messages like axios errors and provides fallbacks
 */
export const extractErrorMessage = (error: any): string => {
  // Default fallback message
  const defaultMessage = 'Something went wrong. Please try again.';

  if (!error) {
    return defaultMessage;
  }

  if (error.errors && error.errors.length > 0) {
    // API error format: { errors: [{ detail: "message" }] }
    return error.errors[0].detail || defaultMessage;
  }

  if (error.detail) {
    // Direct error format: { detail: "message" }
    return error.detail;
  }

  if (error.message && !error.message.includes('status code')) {
    // Only use error.message if it's not an axios error
    return error.message;
  }

  // Fallback for any other case (including axios errors)
  return defaultMessage;
};
