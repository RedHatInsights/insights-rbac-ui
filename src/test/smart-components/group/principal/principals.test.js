import React from 'react';
import { Provider } from 'react-redux';
import { MemoryRouter, Route } from 'react-router-dom';
import configureStore from 'redux-mock-store';
import toJson from 'enzyme-to-json';
import { mock } from '../../../__mocks__/apiMock';
import promiseMiddleware from 'redux-promise-middleware';
import GroupPrincipals from '../../../../smart-components/group/principal/principals';
import { notificationsMiddleware } from '@redhat-cloud-services/frontend-components-notifications';
import { RBAC_API_BASE } from '../../../../utilities/constants';

describe('<GroupPrincipals />', () => {
  const middlewares = [promiseMiddleware, notificationsMiddleware()];
  let mockStore;
  let initialState;

  beforeEach(() => {
    mockStore = configureStore(middlewares);
    initialState = {
      groupReducer: {
        groups: {},
        selectedGroup: {
          members: {
            isLoading: false,
            meta: { count: 2, offset: 0, limit: 10 },
            data: [
              { username: 'test', email: 'test', first_name: 'test', last_name: 'test', is_active: true },
              { username: 'test2', email: 'test2', first_name: 'test2', last_name: 'test2', is_active: false },
            ],
          },
        },
      },
    };
    mock.onGet(`${RBAC_API_BASE}/groups/test-group/principals/`).replyOnce(200);
  });

  it('should render correctly loader', () => {
    const store = mockStore({
      ...initialState,
      groupReducer: {
        ...initialState.groupReducer,
        selectedGroup: {
          members: {
            isLoading: true,
          },
        },
      },
    });
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter initialEntries={['/groups/detail/test-group/members']}>
          <Route path="/groups/detail/:uuid/members" component={GroupPrincipals} />
        </MemoryRouter>
      </Provider>
    );
    expect(toJson(wrapper.find('ListLoader'), { mode: 'shallow' })).toMatchSnapshot();
  });

  it('should render correctly with empty members', () => {
    const store = mockStore({
      ...initialState,
      groupReducer: {
        ...initialState.groupReducer,
        selectedGroup: {
          members: {
            isLoading: false,
            data: [],
            meta: {},
          },
        },
      },
    });
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter initialEntries={['/groups/detail/test-group/members']}>
          <Route path="/groups/detail/:uuid/members" component={GroupPrincipals} />
        </MemoryRouter>
      </Provider>
    );
    expect(toJson(wrapper.find('EmptyWithFilter'), { mode: 'shallow' })).toMatchSnapshot();
  });

  it('should render correctly with org admin rights', () => {
    const store = mockStore({
      groupReducer: {
        ...initialState.groupReducer,
        groups: {
          entitlements: {
            user: {
              is_org_admin: true,
            },
          },
        },
      },
    });
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter initialEntries={['/groups/detail/test-group/members']}>
          <Route path="/groups/detail/:uuid/members" component={GroupPrincipals} />
        </MemoryRouter>
      </Provider>
    );
    expect(toJson(wrapper.find('TableToolbarView'), { mode: 'shallow' })).toMatchSnapshot();
  });

  it('should render correctly with default group', () => {
    const store = mockStore({
      groupReducer: {
        ...initialState.groupReducer,
        selectedGroup: {
          ...initialState.groupReducer.selectedGroup,
          platform_default: true,
          loaded: true,
        },
      },
    });
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter initialEntries={['/groups/detail/test-group/members']}>
          <Route path="/groups/detail/:uuid/members" component={GroupPrincipals} />
        </MemoryRouter>
      </Provider>
    );
    expect(toJson(wrapper.find('#tab-principals'))).toMatchSnapshot();
  });

  it('should render correctly with data', () => {
    const store = mockStore(initialState);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter initialEntries={['/groups/detail/test-group/members']}>
          <Route path="/groups/detail/:uuid/members" component={GroupPrincipals} />
        </MemoryRouter>
      </Provider>
    );
    expect(toJson(wrapper.find('TableToolbarView'), { mode: 'shallow' })).toMatchSnapshot();
  });
});
