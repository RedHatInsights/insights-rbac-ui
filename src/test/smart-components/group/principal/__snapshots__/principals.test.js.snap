// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GroupPrincipals /> should render correctly loader 1`] = `
<ListLoader
  items={5}
/>
`;

exports[`<GroupPrincipals /> should render correctly with data 1`] = `
<TableToolbarView
  actionResolver={[Function]}
  borders={true}
  checkedRows={Array []}
  columns={
    Array [
      Object {
        "title": "Status",
        "transforms": Array [
          [Function],
        ],
      },
      Object {
        "title": "Username",
      },
      Object {
        "title": "Email",
      },
      Object {
        "title": "Last name",
      },
      Object {
        "title": "First name",
      },
    ]
  }
  createRows={[Function]}
  data={
    Array [
      Object {
        "email": "test",
        "first_name": "test",
        "is_active": true,
        "last_name": "test",
        "username": "test",
        "uuid": "test",
      },
      Object {
        "email": "test2",
        "first_name": "test2",
        "is_active": false,
        "last_name": "test2",
        "username": "test2",
        "uuid": "test2",
      },
    ]
  }
  emptyProps={
    Object {
      "description": Array [
        "Add a user to configure user access.",
        "",
      ],
      "title": "There are no members in this group",
    }
  }
  fetchData={[Function]}
  filterItems={Array []}
  filterPlaceholder="username"
  filterValue=""
  filters={Array []}
  hideFilterChips={false}
  isCollapsible={false}
  isCompact={false}
  isFilterable={false}
  isLoading={false}
  isSelectable={false}
  pagination={
    Object {
      "count": 2,
      "limit": 10,
      "offset": 0,
    }
  }
  routes={[Function]}
  rowWrapper={[Function]}
  setCheckedItems={[Function]}
  setFilterValue={[Function]}
  titlePlural="members"
  titleSingular="member"
  toolbarButtons={[Function]}
/>
`;

exports[`<GroupPrincipals /> should render correctly with default group 1`] = `
Array [
  <Section
    id="tab-principals"
    type="content"
  >
    <section
      className="ins-l-content"
      id="tab-principals"
    >
       
      <Card>
        <article
          className="pf-c-card"
          data-ouia-component-id={26}
          data-ouia-component-type="PF4/Card"
          data-ouia-safe={true}
        >
          <CardBody>
            <div
              className="pf-c-card__body"
            >
              <Bullseye>
                <div
                  className="pf-l-bullseye"
                >
                  <TextContent>
                    <div
                      className="pf-c-content"
                    >
                      <Text
                        component="h1"
                      >
                        <h1
                          className=""
                          data-pf-content={true}
                        >
                          All users in this organization are members of this group.
                        </h1>
                      </Text>
                    </div>
                  </TextContent>
                </div>
              </Bullseye>
            </div>
          </CardBody>
        </article>
      </Card>
       
    </section>
  </Section>,
  <section
    className="ins-l-content"
    id="tab-principals"
  >
     
    <Card>
      <article
        className="pf-c-card"
        data-ouia-component-id={26}
        data-ouia-component-type="PF4/Card"
        data-ouia-safe={true}
      >
        <CardBody>
          <div
            className="pf-c-card__body"
          >
            <Bullseye>
              <div
                className="pf-l-bullseye"
              >
                <TextContent>
                  <div
                    className="pf-c-content"
                  >
                    <Text
                      component="h1"
                    >
                      <h1
                        className=""
                        data-pf-content={true}
                      >
                        All users in this organization are members of this group.
                      </h1>
                    </Text>
                  </div>
                </TextContent>
              </div>
            </Bullseye>
          </div>
        </CardBody>
      </article>
    </Card>
     
  </section>,
]
`;

exports[`<GroupPrincipals /> should render correctly with empty members 1`] = `
<EmptyWithFilter
  description={
    Array [
      "Add a user to configure user access.",
      "",
    ]
  }
  icon={[Function]}
  title="There are no members in this group"
/>
`;

exports[`<GroupPrincipals /> should render correctly with org admin rights 1`] = `
<TableToolbarView
  actionResolver={[Function]}
  borders={true}
  checkedRows={Array []}
  columns={
    Array [
      Object {
        "title": "Status",
        "transforms": Array [
          [Function],
        ],
      },
      Object {
        "title": "Username",
      },
      Object {
        "title": "Email",
      },
      Object {
        "title": "Last name",
      },
      Object {
        "title": "First name",
      },
    ]
  }
  createRows={[Function]}
  data={
    Array [
      Object {
        "email": "test",
        "first_name": "test",
        "is_active": true,
        "last_name": "test",
        "username": "test",
        "uuid": "test",
      },
      Object {
        "email": "test2",
        "first_name": "test2",
        "is_active": false,
        "last_name": "test2",
        "username": "test2",
        "uuid": "test2",
      },
    ]
  }
  emptyProps={
    Object {
      "description": Array [
        "Add a user to configure user access.",
        "",
      ],
      "title": "There are no members in this group",
    }
  }
  fetchData={[Function]}
  filterItems={Array []}
  filterPlaceholder="username"
  filterValue=""
  filters={Array []}
  hideFilterChips={false}
  isCollapsible={false}
  isCompact={false}
  isFilterable={false}
  isLoading={false}
  isSelectable={false}
  pagination={
    Object {
      "count": 2,
      "limit": 10,
      "offset": 0,
    }
  }
  routes={[Function]}
  rowWrapper={[Function]}
  setCheckedItems={[Function]}
  setFilterValue={[Function]}
  titlePlural="members"
  titleSingular="member"
  toolbarButtons={[Function]}
/>
`;
